[c10] fix weak_intrusive_ptr lock()

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42944 [WIP][c10d][jit] torchbind ProcessGroup
* #44044 [torchbind] add static define for C++ method
* #44046 [c10d] switch ProcessGroup::Work to be managed by intrusive_ptr
* #44492 [c10] make intrusive_ptr available as a pybind holder type
* #46687 update pybind11 to v2.6.0
* **#46007 [c10] fix weak_intrusive_ptr lock()**

When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

wanchaol committed:
[c10] fix weak_intrusive_ptr lock()

When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

[ghstack-poisoned]

gmagogsfm commented on c10/test/util/intrusive_ptr_test.cpp:
```diff
@@ -1653,6 +1653,15 @@ TEST(WeakIntrusivePtrTest, givenPtr_whenLocking_thenReturnsCorrectObject) {
+  var.ptr.reset();
+  EXPECT_TRUE(var.weak.expired());
+  intrusive_ptr<SomeClass> locked = var.weak.lock();
+  EXPECT_TRUE(!locked.defined());
```
nit: EXPECT_FALSE

gmagogsfm reviewed (approved):
Thanks for catching this!

smessmer commented on c10/util/intrusive_ptr.h:
```diff
@@ -595,7 +595,7 @@ class weak_intrusive_ptr final {
 
   intrusive_ptr<TTarget, NullType> lock() const noexcept {
-    auto refcount = target_->refcount_.load();
+    auto refcount = use_count();
```
Let's rather add an explicit if-test for `NullType::singleton` here because this `load()` call forms a common pattern together with the `while` loop and the `compare_exchange_weak` below. I don't want to break those up.

wanchaol replied:
I was adding a if loop to check `if (expired())` then return `NullType::singleton()`, later I realized the `use_count()` returns `load()` and return 0 for `NullType::singleton()`, which is nice to combine. But yeah if the `load()` explicitly combine with `compare_exchange_weak` to form a pattern, I can revert it back

smessmer commented on c10/test/util/intrusive_ptr_test.cpp:
```diff
@@ -1653,6 +1653,15 @@ TEST(WeakIntrusivePtrTest, givenPtr_whenLocking_thenReturnsCorrectObject) {
   EXPECT_EQ(var.ptr.get(), locked.get());
 }
 
+TEST(WeakIntrusivePtrTest, expiredPtr_whenLocking_thenReturnsNullType) {
```
Test looks good but I don't think it has something to do with the bug you fixed. Afaik, expiring the object doesn't set the weak pointer to null but it still points to the expired object. Let's add a test that has a weak nullptr and tries to lock it.

wanchaol replied:
I think `intrusive_ptr.reset()` will set `target_ = NullType::singleton();`, and weak_intrusive_ptr is actually sharing the same `TTarget*` with intrusive_ptr as it constructs from it? so weak_intrusive_ptr should be null as well. But either way, I added a explicit weak nullptr test to make sure it works :)

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] fix weak_intrusive_ptr lock()"


When owner released the object, target will become null and illegal to
access refcount_ again. This PR fixes this and return null in that case.

Differential Revision: [D24374846](https://our.internmc.facebook.com/intern/diff/D24374846)

[ghstack-poisoned]
